### PR TITLE
Getting closer to something that should work

### PR DESCRIPTION
**Pull request by Yoric**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+            // First run on_failure/on_success.
+            // Store errors for later.
+            let result = match (status, down_script) {
+                (
+                    Status::Failure,
+                    DownScript {
+                        failure: Some(ref on_failure),
+                        ..
+                    },
+                ) => on_failure.run(&*EMPTY_ENV),
```
Comment (Gnuxie) [inline src/lib.rs:219]: Why are we passing an empty environment here?

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+pub fn up(
+    version: SynapseVersion,
+    script: &Option<Script>,
+) -> Result<HashMap<&'static OsStr, OsString>, Error> {
+    // FIXME: Up Synapse.
+    // FIXME: If we have a token for an admin user, test it.
+    // FIXME: Where should we store the token for the admin user? File storage? An embedded db?
+    // FIXME: Note that we need to wait and retry, as bringing up Synapse can take a little time.
+    // FIXME: If we have no token or the token is invalid, create an admin user.
+    // FIXME: If the configuration states that we need to run an `up` script, run it.
```
Comment (Gnuxie) [inline src/lib.rs:197]: if we find that synapse container that is already up are we going to leave it up? Do we keep all of the data it has created if we want to rebuild the image and container?

**Review by Gnuxie — APPROVED** [review]
Looks great :+1:

File: src/lib.rs (hunk 1 above)
Comment (Yoric) [reply src/lib.rs:219]: It's mostly a placeholder because I don't know what we want to pass just yet.

File: src/lib.rs (hunk 2 above)
Comment (Yoric) [reply src/lib.rs:197]: > if we find that synapse container that is already up are we going to leave it up?

I'd say that if is already up, we should leave it up.

> Do we keep all of the data it has created if we want to rebuild the image and container?

Just exactly which data Docker retains has always been a mystery to me. But I don't think we care all that much for the time being.

**Pull request merged**